Add QuadraticHamiltonian class

kevinsung opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Part of #460. Adds a data structure to represent quadratic Hamiltonians.


### Details and comments
I have implemented a method `_fermionic_op` to convert a QuadraticHamiltonian to a FermionicOp. For now, I have kept that method hidden because I'm not sure what general approach we should take to converting between operator types, e.g., whether we want to have instance methods or whether we want to have a functional style like `qiskit_nature.fermionic_op(quadratic_hamiltonian)`

woodsp-ibm commented:
> The CI failure is unrelated to this PR.

Was just fixed so I updated your branch - should pass now

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+        r"""Initialize a QuadraticHamiltonian.
+
+        Args:
+            hermitian_part: The matrix :math:`M` containing the
+                coefficients of the terms that conserve particle number.
```
I think this is line-wrapped too early.
Potentially more below.

kevinsung replied:
I wrapped it earlier; let me know if it's still not okay.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+        else:
+            raise ValueError("Either a Hermitian part or antisymmetric part must be specified.")
+
+        # TODO maybe validate matrix properties
```
Please address this

kevinsung replied:
Done.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+            self._n_orbitals, _ = hermitian_part.shape
+        elif antisymmetric_part is not None:
+            self._n_orbitals, _ = antisymmetric_part.shape
+        else:
+            raise ValueError("Either a Hermitian part or antisymmetric part must be specified.")
```
Would it make sense to provide a keyword argument like `num_orbitals` to handle the creation of a zero-operator?

kevinsung replied:
Done, I named it `num_modes`.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+        """The constant."""
+        return self._constant
+
+    def _fermionic_op(self) -> FermionicOp:
```
Why is this a private method?

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+    def _fermionic_op(self) -> FermionicOp:
+        """Convert to FermionicOp."""
+        terms = []
+        # TODO I shouldn't have to use string labels
```
You can refactor this now that #506 is merged.

kevinsung replied:
Done.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+        """Convert to FermionicOp."""
+        terms = []
+        # TODO I shouldn't have to use string labels
+        # diagonal terms
+        for i in range(self._n_orbitals):
+            terms.append((f"+_{i} -_{i}", self.hermitian_part[i, i]))
+        # off-diagonal terms
+        for i in range(self._n_orbitals):
+            for j in range(i + 1, self._n_orbitals):
```
```suggestion
        for i in range(self._n_orbitals):
            # diagonal terms
            terms.append((f"+_{i} -_{i}", self.hermitian_part[i, i]))
            # off-diagonal terms
            for j in range(i + 1, self._n_orbitals):
```

kevinsung replied:
Done.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+                terms.append((f"+_{j} -_{i}", self.hermitian_part[j, i]))
+                terms.append((f"+_{i} +_{j}", self.antisymmetric_part[i, j]))
+                terms.append((f"-_{j} -_{i}", self.antisymmetric_part[i, j].conj()))
+        # constant
+        terms.append(("", self.constant))
```
How about you initialize `terms = [("", self.constant)]` before entering the loop above?

kevinsung replied:
Done.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+        majorana_basis = np.block([[eye, eye], [1j * eye, -1j * eye]]) / np.sqrt(2)
+        matrix = -1j * majorana_basis.conj() @ original @ majorana_basis.T.conj()
+        constant = 0.5 * np.trace(self.hermitian_part) + self.constant
+        # imaginary parts should be zero
```
Do you want to assert this and raise an error if not?

kevinsung replied:
It's not necessary because it's guaranteed if the input matrices are valid, and if the user turned validation off, then this shouldn't be checked.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+        (which satisfies additional constraints).
+
+        Returns:
+            - The matrix :math:`W`, which is either an :math:`N \times N` or
+              an :math:`N \times 2N` matrix
```
Maybe repeat that the latter is the case the `\Delta != 0`

kevinsung replied:
Done.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,306 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
```
I believe that, because of the way how our copyright checker works, this is going to fail once merged to `main` because on that branch a single squashed commit dated to 2022 will be the only commit history related to this file.

@manoelmarques is there anything reasonable we can do about that?

woodsp-ibm replied:
I did not believe we checked the validity as such of the earliest date the code was copyrighted - unless that's the only date in there since its the same as the last date it was updated, due to code getting moved around across repos/refactored etc its kinda difficult to do. The time the code was last updated, the latter (or sole) date, we did check to make sure it was current with the changes.

mrossinek reviewed (commented):
As outlined [here](https://github.com/Qiskit/qiskit-nature/issues/460#issuecomment-1039226596) we can proceed with this PR.

kevinsung commented:
@mrossinek I have addressed all your comments besides the one related to the copyright checker. If needed, I can manually squash my commits into a single one with a 2022 timestamp.

mrossinek commented:
> I have addressed all your comments besides the one related to the copyright checker. If needed, I can manually squash my commits into a single one with a 2022 timestamp.

Sounds like the copyright should be fine as only the last year is being checked. Please fix CI and then we can progress with merging this :+1:

kevinsung commented:
@mrossinek I'm still issues running checks locally (see https://github.com/Qiskit/qiskit-nature/issues/567) but I've fixed the issues raised by CI.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,349 @@
+        """Whether the Hamiltonian conserves particle number."""
+        return np.allclose(self.antisymmetric_part, 0.0)
+
+    def majorana_form(self) -> tuple[np.ndarray, float]:
```
In order to be able to use a type hint such as this one in Python versions <3.9 you have to add the following import:
```python
from __future__ import annotations
```
That is what is causing the CI to break.

kevinsung replied:
Done.

kevinsung commented:
@woodsp-ibm I've added a release note.

woodsp-ibm commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,351 @@
+        if self._antisymmetric_part is None:
+            self._antisymmetric_part = np.zeros((self._num_modes, self._num_modes))
+
+    @property
+    def hermitian_part(self) -> np.ndarray:
+        """The matrix of coefficients of terms that conserve particle number."""
+        return self._hermitian_part
+
+    @property
+    def antisymmetric_part(self) -> np.ndarray:
+        """The matrix of coefficients of terms that do not conserve particle number."""
+        return self._antisymmetric_part
+
```
So this has a number of getters but no setters to change this instance later?

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,351 @@
+        """The number of modes this operator acts on."""
+        return self._num_modes
+
+    def to_fermionic_op(self) -> FermionicOp:
```
I am wondering if we should rename this to `to_second_q_op` which is inline with the namings in the `properties` module.
If we (in the future) design a common base class for some matrix container that will likely be the name of this function, avoiding the deprecation of this one.

kevinsung replied:
I think the current name is much better because it precisely describes what the method does. `to_second_q_op` is confusing because QuadraticHamiltonian is already a "second-quantized op" (though not a SecondQuantizedOp). From the perspective of a user trying to discover this method, I think they are more likely to guess the current name.

mrossinek commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,351 @@
+
+    def to_fermionic_op(self) -> FermionicOp:
+        """Convert to FermionicOp."""
+        terms = [([], self.constant)]  # type: list[tuple[list[tuple[str, int]], complex]]
```
```suggestion
        terms: list[tuple[list[tuple[str, int]], complex]] = [([], self.constant)]
```

kevinsung replied:
Done.

mrossinek reviewed (commented):
Just a minor comment, other than that this lgtm :+1:

woodsp-ibm commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,351 @@
+        rtol: float = 1e-5,
+        atol: float = 1e-8,
+    ) -> None:
+        r"""Initialize a QuadraticHamiltonian.
```
This text can/should be removed `Initialize a QuadraticHamiltonian.` We had other places where things like X constructor is put as text. Since this is Python this is what the init method its not necessary to state the obvious. As the class docstring is merged with the docstring here when its comes out in the docs we can go straight to args I think

![image](https://user-images.githubusercontent.com/40241007/161825178-35b9f2da-51bb-426d-8c01-18af45d10649.png)

kevinsung replied:
Done.

woodsp-ibm commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,351 @@
+            constant: An additive constant term.
+            num_modes: Number of fermionic modes. This should be consistent with hermitian_part
+                and antisymmetric_part if they are specified.
+            validate: Whether to validate the inputs.
```
What is the rationale of giving a user the choice here (performance over hermitian and antisymmetric checks?)? I imagine things will go bad if the inputs do not conform to certain expectations, though I see these checks are only done when both matrices are specified. Things like the hermitian and anti-symmetric tests are not done on individual inputs. Nor is num_modes checked, if given, if only a single matrix is supplied.

kevinsung replied:
Yes, the user should be allowed to turn off validation for performance reasons.

>Things like the hermitian and anti-symmetric tests are not done on individual inputs. Nor is num_modes checked, if given, if only a single matrix is supplied.

This is incorrect. The checks are, in fact, performed in those situations, and I have included tests for that behavior.

woodsp-ibm replied:
My bad, I read the first if, under validate, of the and of the two and somehow misread the other tests as being nested under that. I see now that they are at the same level,

woodsp-ibm commented on qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py:
```diff
@@ -0,0 +1,351 @@
+                and antisymmetric_part is not None
+                and hermitian_part.shape != antisymmetric_part.shape
+            ):
+                raise ValueError("Hermitian part and antisymmetric part must have same shape.")
```
In these error messages it is helpful for a user and debugging etc for them to be formatted with the actual failing values.

kevinsung replied:
Done.

kevinsung commented:
@woodsp-ibm I have addressed your comments. Please take another look.

woodsp-ibm commented:
I think you missed the **f** from the string they now need to be like `f"my value {x}"` for the vars to be formatted in.

Also although I put the comment in one place for the error message I was meaning it could apply to a number of the error messages in the various checks to show the failing values where that's applicable.
